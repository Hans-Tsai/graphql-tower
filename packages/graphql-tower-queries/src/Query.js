import _ from 'lodash';
import PayloadField from './PayloadField';
import GraphQLField from './GraphQLField';

export default class Query extends GraphQLField {
  static resolveArgs(...args) {
    return _.defaultsDeep(_.mapValues(this._.parentArgs, (value, key) => (
      value instanceof PayloadField ? value.resolveValue(...args, key) : value
    )), args[1]);
  }

  static async preResolve(payload, client, context, info) {
    const { resolveArgs, middleware, resolve } = this.constructor;
    const args = resolveArgs.call(this, payload, client, context, info);
    await middleware.call(this, payload, args, context, info);
    return resolve.call(this, payload, args, context, info);
  }

  constructor(parentArgs) {
    super();

    this._.parentArgs = parentArgs;

    Object.defineProperty(this, 'resolve', {
      enumerable: true,
      set: (resolve) => { this._.resolve = resolve; },
      get: () => this.constructor.preResolve.bind(this),
    });
  }
}
