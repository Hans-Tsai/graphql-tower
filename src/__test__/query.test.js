import _ from 'lodash';
import faker from 'faker';
import { GraphQLInt, GraphQLList } from 'graphql';
import Query, { QueryWithConnection } from '../query';

describe('query', () => {
  it('Query', async () => {
    const data = { id: faker.random.number() };
    const resolve = jest.fn(() => Promise.resolve(data));

    const QueryNode = class extends Query {};
    const queryNode = new QueryNode();
    expect(queryNode).toMatchSnapshot();
    expect(await queryNode.resolve()).toBeUndefined();

    const QueryResolve = class extends Query {
      resolve = resolve;
    };
    const queryResolve = new QueryResolve();
    resolve.mockClear();
    expect(await queryResolve.resolve()).toEqual(data);
    expect(resolve).toHaveBeenCalledTimes(1);

    const queryExtend = _.extend({}, queryResolve);
    expect(queryExtend.resolve).not.toBeUndefined();
  });

  it('Query setup middleware', async () => {
    const data = { id: faker.random.number() };
    const resolve = jest.fn(() => Promise.resolve(data));
    const middleware = jest.fn();

    const QueryMiddleware = class extends Query {
      middleware = middleware;
      resolve = resolve;
    };
    const queryMiddleware = new QueryMiddleware();

    middleware.mockClear();
    resolve.mockClear();
    expect(await queryMiddleware.resolve()).toEqual(data);
    expect(middleware).toHaveBeenCalledTimes(1);
    expect(resolve).toHaveBeenCalledTimes(1);

    const middlewarePromise = [];
    const QueryDeepMiddleware = class extends QueryMiddleware {
      middleware = [
        () => new Promise(done => setImmediate(() => {
          middlewarePromise.push(1);
          done();
        })),
        () => (middlewarePromise.push(2)),
      ];
      resolve = resolve;
    };
    const queryDeepMiddleware = new QueryDeepMiddleware();

    middleware.mockClear();
    resolve.mockClear();
    expect(await queryDeepMiddleware.resolve()).toEqual(data);
    expect(middleware).toHaveBeenCalledTimes(1);
    expect(resolve).toHaveBeenCalledTimes(1);
    expect(middlewarePromise).toEqual([1, 2]);

    expect(() => {
      const QueryFailed = class extends Query {
        middleware = faker.lorem.word();
        resolve = resolve;
      };
      const queryFailed = new QueryFailed();
      queryFailed.resolve();
    }).toThrowError('middleware a function array is required');
  });

  it('Query setup afterware', async () => {
    const data = { id: faker.random.number() };
    const reply = { id: faker.random.number() };
    const resolve = jest.fn(() => Promise.resolve(data));
    const afterware = jest.fn(() => Promise.resolve(reply));

    const QueryAfterware = class extends Query {
      afterware = afterware;
      resolve = resolve;
    };
    const queryAfterware = new QueryAfterware();

    afterware.mockClear();
    resolve.mockClear();
    expect(await queryAfterware.resolve()).toEqual(reply);
    expect(afterware).toHaveBeenCalledTimes(1);
    expect(afterware).toHaveBeenLastCalledWith(undefined, undefined, undefined, undefined, data);
    expect(resolve).toHaveBeenCalledTimes(1);

    const QueryDeepAfterware = class extends QueryAfterware {
      afterware = [
        (payload, args, context, info, results) => new Promise(done => setImmediate(() => {
          done([results, 1]);
        })),
        (payload, args, context, info, results) => _.flatten([results, 2]),
      ];;
      resolve = resolve;
    };
    const queryDeepAfterware = new QueryDeepAfterware();

    afterware.mockClear();
    resolve.mockClear();
    expect(await queryDeepAfterware.resolve()).toEqual([reply, 1, 2]);
    expect(afterware).toHaveBeenCalledTimes(1);
    expect(afterware).toHaveBeenLastCalledWith(undefined, undefined, undefined, undefined, data);
    expect(resolve).toHaveBeenCalledTimes(1);

    expect(() => {
      const QueryFailed = class extends Query {
        afterware = faker.lorem.word();
        resolve = resolve;
      };
      const queryFailed = new QueryFailed();
      queryFailed.resolve();
    }).toThrowError('afterware a function array is required');
  });

  it('QueryWithConnection', () => {
    const QueryConnection = class extends QueryWithConnection {
      type = GraphQLInt;
    };
    const query = new QueryConnection();
    expect(query).toMatchSnapshot();

    const queryExtend = _.extend({}, query);
    expect(queryExtend.type).not.toBeUndefined();
    expect(queryExtend.resolve).not.toBeUndefined();
  });
});
