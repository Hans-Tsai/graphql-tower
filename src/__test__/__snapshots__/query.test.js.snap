// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`query Query 1`] = `
QueryNode {
  "_": Object {
    "afterware": Array [],
    "middleware": Array [],
    "parentArgs": undefined,
    "resolve": [Function],
  },
  "args": Object {},
  "resolve": [Function],
}
`;

exports[`query QueryWithConnection 1`] = `
QueryConnection {
  "_": Object {
    "afterware": Array [],
    "middleware": Array [],
    "node": "Int",
    "parentArgs": undefined,
    "resolve": [Function],
    "type": "[Int]",
  },
  "args": Object {
    "after": Object {
      "type": "String",
    },
    "first": Object {
      "type": "Int",
    },
    "offset": Object {
      "type": "Int",
    },
  },
  "resolve": [Function],
  "type": "[Int]",
}
`;

exports[`query QueryWithNode 1`] = `
QueryNode {
  "_": Object {
    "afterware": Array [],
    "middleware": Array [],
    "parentArgs": Object {
      "id": PayloadField {
        "resolveValue": [Function],
      },
    },
    "resolve": [Function],
  },
  "args": Object {},
  "fieldName": undefined,
  "resolve": [Function],
  "type": "Int",
}
`;
